mllib/src/main/scala/org/apache/spark/ml/optim/NormalEquationSolver.scala

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.ml.optim
+
+import breeze.linalg.{DenseVector => BDV}
+import breeze.optimize.{CachedDiffFunction, DiffFunction, LBFGS => BreezeLBFGS, OWLQN => BreezeOWLQN}
+import scala.collection.mutable
+
+import org.apache.spark.ml.linalg.{BLAS, DenseVector, Vectors}
+import org.apache.spark.mllib.linalg.CholeskyDecomposition
+
+private[ml] class NormalEquationSolution(
+    val fitIntercept: Boolean,
+    private val _coefficients: Array[Double],
+    val aaInv: Option[DenseVector],
```

Comment:
I'd use ```Option[Array[Double]]``` to represent ```aaInv```, since we will scale it to original space at ```WeightedLeastSquares``` L263.
BTW, should we use also expose ```coefficients``` as ```Array[Double]``` rather than ```DenseVector```?
I found we did in-place update to scale back at ```WeightedLeastSquares``` L255. I'm more prefer to get coefficients in array by solver, then scale back and wrap it with ```DenseVector``` at upper layer. You can refer the ```LinearRegression``` solved by ```L-BFGS```. Only suggestion, not big deal.